src/py/crankshaft/crankshaft/clustering/kmeans.py

```diff
@@ -1,18 +1,98 @@
 from sklearn.cluster import KMeans
 import plpy
+import numpy as np
+
 
 def kmeans(query, no_clusters, no_init=20):
-    data = plpy.execute('''select array_agg(cartodb_id order by cartodb_id) as ids,
-        array_agg(ST_X(the_geom) order by cartodb_id) xs,
-        array_agg(ST_Y(the_geom) order by cartodb_id) ys from ({query}) a
-        where the_geom is not null
-    '''.format(query=query))
-
-    xs  = data[0]['xs']
-    ys  = data[0]['ys']
+    """
+        find centers based on clusteres of latitude/longitude pairs
+        query: SQL query that has a WGS84 geometry (the_geom)
+    """
+    full_query = ("SELECT array_agg(cartodb_id ORDER BY cartodb_id) as ids,"
+                  "array_agg(ST_X(the_geom) ORDER BY cartodb_id) xs,"
+                  "array_agg(ST_Y(the_geom) ORDER BY cartodb_id) ys "
+                  "FROM ({query}) As a "
+                  "WHERE the_geom IS NOT NULL").format(query=query)
+    try:
+        data = plpy.execute(full_query)
+    except plpy.SPIError, err:
+        plpy.error("k-means (spatial) cluster analysis failed: %s" % err)
+
+    # Unpack query response
+    xs = data[0]['xs']
+    ys = data[0]['ys']
     ids = data[0]['ids']
 
-    km = KMeans(n_clusters= no_clusters, n_init=no_init)
-    labels = km.fit_predict(zip(xs,ys))
-    return zip(ids,labels)
+    km = KMeans(n_clusters=no_clusters, n_init=no_init)
+    labels = km.fit_predict(zip(xs, ys))
+    return zip(ids, labels)
+
+
+def kmeans_nonspatial(query, colnames, num_clusters=5,
```

Comment:
It might also be worthwhile calculating a [Silhouette](https://en.wikipedia.org/wiki/Silhouette_(clustering)) score for each cluster and returning that as well. This would give us a metric of how good each cluster is. You can do this relatively easily in sklearn : http://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html